### ISSUE TITLE
Support for user-defined chunk_alloc_mmap/chunk_alloc_munmap

### ISSUE THREAD

**Issue by krzycz** [jemalloc/jemalloc#201]
User defined chunk allocators have no access to the extant/free chunk trees, so they can't use the *jemalloc* built-in chunk recycling logic. Moreover, when implementing such functionality in the custom chunk allocator, it's not even possible to re-use some part of *jemalloc* implementation (except by duplicating some *jemalloc* code), as it's all internal.

For some applications however, all what the custom allocator is expected to do is to add some additional flags to `mmap()` call (like `MAP_HUGETLB`), set the NUMA memory policy, etc. So, in many cases, instead of replacing the entire `chunk_alloc` path with the custom implementation, it would be enough to substitute only the bottom part of `chunk_alloc` (I mean `chunk_alloc_mmap/chunk_alloc_dss`) with some user-defined function. This would reduce the amount of code required to implement the custom chunk allocator, and it would also allow the custom allocator could still benefit from the chunk recycling mechanism.

Initially, I was thinking about adding an optional third function to the existing `chunk_alloc_mmap` and `chunk_alloc_dss`, along with its own dedicated set of chunk trees. But after all, I think that it will be easier to allow user to simply override the default implementation of `chunk_alloc_mmap` with its own version. The custom function could then reuse the existing `chunks_szad_mmap/chunks_ad_mmap` trees.
Also, since the trees are now per arena, then chunks mapped using the custom function would be isolated from the chunks coming from the default *jemalloc* function.

I believe such functionality won't be in conflict with the existing user-defined `chunk_alloc/chunk_dalloc` feature. It's still up to the user if their prefer to override the entire `chunk_alloc` stuff, or just the bottom part of it.
Seems like it would also address the issue #176.

The prototype implementation is available here: https://github.com/krzycz/jemalloc/commit/eb5364b282eb857fa2e6d00df8872356557962bc

Please, take a look and let me know what do you think about it.

**Comment by jasone** [comment]
I'm conflicted about this approach, because in order to be effectively used, proper dss configuration is also necessary.  In general I'm not very happy with how mmap/dss precedence configuration works...  Can you think of any ways we could simplify one or both interfaces such that interactions aren't surprising?

**Comment by krzycz** [comment]
The prototype is based on the fact that dss precedence is by default "secondary", so it's enough to override `chunk_alloc_mmap` to switch to the user-defined function.
I agree that in general, the user must take care about proper dss configuration. Probably, to avoid any interactions, the best option is to disable dss for given arena if custom `chunk_mmap` is set. This could be done automatically in `arena_i_chunk_mmap_ctl`, or we can modify `chunk_alloc_core` to ignore dss completely, if `arena->chunk_mmap` points to the user-defined function.
Do you think it would be enough?

We can also go further and change the way how the dss/mmap(/custom) precedence configuration works. For instance, for each arena we can have an array of two or three pointers to `chunk_alloc_mmap`, `chunk_alloc_dss` or user-defined function (same for dalloc). By setting them in a proper order, we could configure the desired precedence. And by setting some of those pointers to NULL, we could disable selected paths for given arena. I.e.

- default configuration:
 - chunk_alloc_mmap
 - chunk_alloc_dss
 - NULL

- user-defined function only:
 - chunk_alloc_user
 - NULL
 - NULL

- primary - user, secondary - mmap:
 - chunk_alloc_user
 - chunk_alloc_mmap
 - NULL

and so on...

But this would still require to set both the pointer to user-defined function and the precedence config. (Unless `mallctl` for `arena.<i>.dss` does not take the pointer to such array, allowing to provide both the precedence and the user-defined function pointer in a single `mallctl` call, and thus eliminating `arena.<i>.chunk.mmap`.)

**Comment by cmcantalupo** [comment]
In the memkind fork of jemalloc from 3.5.1 it was required that we partition the red-black tree used for the huge allocations to avoid user specialized chunks from getting mixed.  Does the new radix tree implementation for huge allocations now in the dev branch have the same requirement for partitioning?

**Comment by jasone** [comment]
@krzycz, that's good food for thought; thanks.

@cmcantalupo, the radix tree contains one entry for each extant huge allocation which is used to find the corresponding ```extent_node_t```, regardless of whether allocated via dss, mmap, or an application-provided function.  Internally, there remain three separate pairs of red-black trees for dss, mmap, and cached chunks, because of the desire to search them in a particular order.  Does that answer your question, or did I miss the point?

**Comment by cmcantalupo** [comment]
Is it correct that the extent, address and radix trees all share the same nodes (where each node contains the edges for all three trees)?  It it also true that the radix tree is used for insertion and deletion, but the red-black trees are used for search?  If this the case then it seems the memkind red-black tree partitioning extension can be adapted to the new jemalloc huge extent implementation because the partitioning is only relevant to the search operation.  

The goal of this partitioning is to enable the selection of OS memory policies from a heap management interface.  This means that we want to be able to distinguish virtual address ranges which have different OS policies applied to them.

**Comment by cmcantalupo** [comment]
@jasone, Sorry that I may have confused this issue more than I clarified.  I will review the radix tree implementation and get back here about how we may keep virtual address policies separate for huge allocations.  I think it should be straight forward.  

I have posted an article that we submitted to Runtime and Operating Systems for Supercomputers (ROSS) 2015 here:

http://memkind.github.io/memkind/memkind_arch_20150318.pdf

This may serve to motivate the use case for the feature described in this issue, and give you some more background on the approach we are taking.  Thanks for trying to clarify the implementation details.
